src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -248,12 +275,34 @@ protected AnalyzerDriver(ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerM
         /// </summary>
         public async Task<ImmutableArray<Diagnostic>> GetDiagnosticsAsync()
         {
-            var allDiagnostics = DiagnosticBag.GetInstance();
             if (CompilationEventQueue.IsCompleted)
             {
                 await this.WhenCompletedTask.ConfigureAwait(false);
             }
 
+            return GetDiagnostics();
+        }
+
+        /// <summary>
+        /// Returns all diagnostics produced by a partial analysis of a single syntax tree.
+        /// </summary>
+        public async Task<ImmutableArray<Diagnostic>> GetPartialDiagnosticsAsync(SyntaxTree syntaxTree, CancellationToken cancellationToken)
+        {
+            if (_initializeTaskStarted && _primaryTask == null)
+            {
+                await _initializeTask.ConfigureAwait(false);
+                await ExecuteSyntaxTreeActions(ImmutableArray.Create(syntaxTree), cancellationToken).ConfigureAwait(false);
```

Comment:
You've understood it well enough. You're right--this needs to be protected against concurrent requests.